internal/pally/registry.go

```diff
@@ -0,0 +1,250 @@
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+	c := newCounter(opts)
+	if err := r.register(c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// MustCounter constructs a new Counter. It panics if it encounters an error.
+func (r *Registry) MustCounter(opts Opts) Counter {
+	c, err := r.NewCounter(opts)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create Counter with options %+v: %v", opts, err))
+	}
+	return c
+}
+
+// NewGauge constructs a new Gauge.
+func (r *Registry) NewGauge(opts Opts) (Gauge, error) {
+	opts = r.addConstLabels(opts)
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+	g := newGauge(opts)
+	if err := r.register(g); err != nil {
+		return nil, err
+	}
+	return g, nil
+}
+
+// MustGauge constructs a new Gauge. It panics if it encounters an error.
+func (r *Registry) MustGauge(opts Opts) Gauge {
+	g, err := r.NewGauge(opts)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create Gauge with options %+v: %v", opts, err))
+	}
+	return g
+}
+
+// NewCounterVector constructs a new CounterVector.
+func (r *Registry) NewCounterVector(opts Opts) (CounterVector, error) {
+	opts = r.addConstLabels(opts)
+	if err := opts.validateVector(); err != nil {
+		return nil, err
+	}
+	v := newCounterVector(opts)
+	if err := r.register(v); err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// MustCounterVector constructs a new CounterVector. It panics if it encounters
+// an error.
+func (r *Registry) MustCounterVector(opts Opts) CounterVector {
+	v, err := r.NewCounterVector(opts)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create CounterVector with options %+v: %v", opts, err))
+	}
+	return v
+}
+
+// NewGaugeVector constructs a new CounterVector.
+func (r *Registry) NewGaugeVector(opts Opts) (GaugeVector, error) {
+	opts = r.addConstLabels(opts)
+	if err := opts.validateVector(); err != nil {
+		return nil, err
+	}
+	v := newGaugeVector(opts)
+	if err := r.register(v); err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// MustGaugeVector constructs a new GaugeVector. It panics if it encounters an
+// error.
+func (r *Registry) MustGaugeVector(opts Opts) GaugeVector {
+	v, err := r.NewGaugeVector(opts)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create GaugeVector with options %+v: %v", opts, err))
+	}
+	return v
+}
+
+func (r *Registry) exportToTally() {
+	defer close(r.stopped)
+	// When stopping, do one last export to catch any stragglers.
+	defer r.push()
+
+	for {
+		select {
+		case <-r.stop:
+			return
+		case <-r.ticker.C:
+			r.push()
+		}
+	}
+}
+
+func (r *Registry) push() {
+	r.metricsMu.RLock()
+	for _, m := range r.metrics {
+		m.push(r.scope)
+	}
+	r.metricsMu.RUnlock()
+}
+
+func (r *Registry) register(m metric) error {
+	r.metricsMu.Lock()
+	r.metrics = append(r.metrics, m)
+	r.metricsMu.Unlock()
+
+	if err := r.prom.Register(m); err != nil {
+		return err
+	}
+	for _, fed := range r.federated {
+		if err := fed.Register(m); err != nil {
+			return err
```

Comment:
If a metric fails to register fully, we'll still be emitting to it since it's in `r.metrics`.

Is the expectation that the program will die if it fails to register metrics so invalid state is okay?